Reorganize host.rs and wasm32.rs.

sunfishcode opened the pull request:
Reorganize host.rs and wasm32.rs into host.rs, wasi.rs, and wasi32.rs.
    
Most of the contents of host.rs was not actually host-specific, as most
of the types are fixed-size types like u32 or i64. These types are now
in wasi.rs.
    
The few types which do have pointer or usize-sized values now remain,
in two versions: host.rs has versions which use actual raw pointers and
usize, and wasi32.rs has versions which use u32 to represent them.

I'm not sure this is the right approach, and welcome comments here!

This approach does not provide a type-level distinction between most
host types and wasm types, though this isn't new; the previous code
also didn't make a strong distinction, because eg. `host::__wasi_filetype_t`
was the same type as `wasm32::__wasi_filetype_t`.

sunfishcode committed:
Reorganize host.rs and wasm32.rs.

Reorganize host.rs and wasm32.rs into host.rs, wasi.rs, and wasi32.rs.

Most of the contents of host.rs was not actually host-specific, as most
of the types are fixed-size types like u32 or i64. These types are now
in wasi.rs.

The few types which do have pointer or usize-sized values now remain,
in two versions: host.rs has versions which use actual raw pointers and
usize, and wasi32.rs has versions which use u32 to represent them.

kubkon reviewed (approved):
This looks great, thanks @sunfishcode! I especially like the fact that we're already factoring in the fact that in the future we expect 64bit WASI.

Commit:
Fix compilation on BSD

Commit:
Fix compilation on Windows

marmistrz reviewed (approved):
Sounds great! I really like the fact that those unclear cases need to be explicitly opted-in by using either host or wasi32.

One nitpick: how about removing `use crate::wasi::*` from `wasi32.rs` and `host.rs` so that `wasi.rs` and X are disjoint and that there's only one way of using the cross-platform types?

It would probably also be a good idea to add some comment about what the difference between `host` and `wasi32` actually is.

kubkon commented:
> Sounds great! I really like the fact that those unclear cases need to be explicitly opted-in by using either host or wasi32.
> 
> One nitpick: how about removing `use crate::wasi::*` from `wasi32.rs` and `host.rs` so that `wasi.rs` and X are disjoint and that there's only one way of using the cross-platform types?
> 
> It would probably also be a good idea to add some comment about what the difference between `host` and `wasi32` actually is.

Hmm, I think I actually prefer the way it is now, since, as far as I understand, for the lib user it is now a matter of importing either `wasi_common::wasi32` or `wasi_common::wasi64` as both (would) include the common `wasi` module, rather than importing `wasi_common::wasi32` and `wasi_common::wasi` separately. A matter of taste I guess ;-)

When it comes to having `crate::wasi::*` inside of the `host` module, I'm fine with excluding the import here as the `host` module should only be used for internal purposes (I think).

kubkon commented:
> > Sounds great! I really like the fact that those unclear cases need to be explicitly opted-in by using either host or wasi32.
> > One nitpick: how about removing `use crate::wasi::*` from `wasi32.rs` and `host.rs` so that `wasi.rs` and X are disjoint and that there's only one way of using the cross-platform types?
> > It would probably also be a good idea to add some comment about what the difference between `host` and `wasi32` actually is.
> 
> Hmm, I think I actually prefer the way it is now, since, as far as I understand, for the lib user it is now a matter of importing either `wasi_common::wasi32` or `wasi_common::wasi64` as both (would) include the common `wasi` module, rather than importing `wasi_common::wasi32` and `wasi_common::wasi` separately. A matter of taste I guess ;-)
> 
> When it comes to having `crate::wasi::*` inside of the `host` module, I'm fine with excluding the import here as the `host` module should only be used for internal purposes (I think).

Actually, I misread this. Unless we make `use crate::wasi::*` public, I agree with @marmistrz that it might make more sense to make `wasi32` and `host` disjoint by removing the import altogether.

kubkon commented on src/wasi32.rs:
```diff
@@ -0,0 +1,232 @@
+pub(crate) const PTRDIFF_MAX: u32 = 2147483647;
+pub(crate) const SIG_ATOMIC_MIN: i32 = -2147483648;
+pub(crate) const SIG_ATOMIC_MAX: u32 = 2147483647;
+pub(crate) const SIZE_MAX: u32 = 4294967295;
```
Yet another thing that slipped through the radar, shouldn't these be public? I'm not sure whether all of them, but type aliases such as `uintptr_t` for sure as it's needed in order to plug in the lib into the runtime, for instance, as is the case in [wasmtime_wasi#L166].

[wasmtime_wasi#L166]: https://github.com/CraneStation/wasmtime/blob/master/wasmtime-wasi/src/syscalls.rs#L166

jedisct1 commented:
The `wasm` and `host` versions of fixed-size types had the advantage of making a distinction between what has to be encoded as little-endian vs what may use a different encoding.

sunfishcode commented:
> One nitpick: how about removing use crate::wasi::* from wasi32.rs and host.rs so that wasi.rs and X are disjoint and that there's only one way of using the cross-platform types?

It a plain `use`, not a `pub use`, so iirc it's not exported. host.rs and wasi32.rs do depend on things declared in wasi.rs, so they need the import, but they don't export it to their users. There's only one way to use the cross-platforms types, and that's `wasi::`.

> Yet another thing that slipped through the radar, shouldn't these be public? I'm not sure whether all of them, but type aliases such as uintptr_t for sure as it's needed in order to plug in the lib into the runtime, for instance, as is the case in wasmtime_wasi#L166.

> The wasm and host versions of fixed-size types had the advantage of making a distinction between what has to be encoded as little-endian vs what may use a different encoding.

There are a few different ways to could approach this, but I'd like to propose the following approach:

Types like `__wasi_fd_t` are always in host-endian order. The only place we should have endianness conversions happening is in src/memory.rs when we're loading out of a wasm linear memory (which is explicitly little-endian), and while this isn't the case yet, it shouldn't cast into `__wasi_fd_t` and other types until it's done the conversion. Hopefully soon, most of src/memory.rs will be auto-generated, which should make this feasible.

Does that make sense?

sunfishcode commented:
Oops, forgot to actually reply to this:
> Yet another thing that slipped through the radar, shouldn't these be public? I'm not sure whether all of them, but type aliases such as uintptr_t for sure as it's needed in order to plug in the lib into the runtime, for instance, as is the case in wasmtime_wasi#L166.

I believe this is sorted out in the rebase I just pushed.

sunfishcode committed:
Fully encapsulate endianness in memory.rs.

This refactors memory.rs to fully encapsulte endianness concerns, so
that outside that file, all values are in host-endian order.

This adds a dependency on the `num` crate, though it's only used for
the `PrimInt` trait, for handling endianness in a generic way.

sunfishcode commented:
With the latest push, I've now added a patch implementing [this idea](https://github.com/CraneStation/wasi-common/pull/151#issuecomment-547651359), of fully encapsulating all endianness issues in memory.rs.

I haven't tested it on a big-endian platform, so it's possible there are endianness bugs (though the code before this patch definitely has endianness bugs), however now we have a principled and simple approach, and we can easily fix any bugs that are found.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
-    pub(crate) u: __wasi_prestat_t___wasi_prestat_u,
+pub struct __wasi_prestat_t {
+    pub pr_type: __wasi_preopentype_t,
+    pub u: __wasi_prestat_t___wasi_prestat_u,
```
Is there a reason why these are `pub` and not `pub(crate)`? AFAICS, `mod host` is private in `lib.rs`.

sunfishcode replied:
Indeed, I've now made these all be `pub(crate)`.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
-pub(crate) union __wasi_prestat_t___wasi_prestat_u {
-    pub(crate) dir: __wasi_prestat_t___wasi_prestat_u___wasi_prestat_u_dir_t,
+pub union __wasi_prestat_t___wasi_prestat_u {
+    pub dir: __wasi_prestat_t___wasi_prestat_u___wasi_prestat_u_dir_t,
```
Same here.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
-pub(crate) struct __wasi_prestat_t___wasi_prestat_u___wasi_prestat_u_dir_t {
-    pub(crate) pr_name_len: usize,
+pub struct __wasi_prestat_t___wasi_prestat_u___wasi_prestat_u_dir_t {
+    pub pr_name_len: usize,
```
Same here.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
-    pub(crate) fs_rights_inheriting: __wasi_rights_t,
+pub struct __wasi_ciovec_t {
+    pub buf: *const void,
+    pub buf_len: usize,
```
Same here.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
-    pub(crate) fd: __wasi_fd_t,
+pub struct __wasi_iovec_t {
+    pub buf: *mut void,
+    pub buf_len: usize,
```
Same here.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
 
 #[allow(unused)]
-pub(crate) unsafe fn ciovec_to_host(ciovec: &__wasi_ciovec_t) -> io::IoSlice {
+pub unsafe fn ciovec_to_host(ciovec: &__wasi_ciovec_t) -> io::IoSlice {
```
Same here.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
 
 #[allow(unused)]
-pub(crate) unsafe fn ciovec_to_host_mut(ciovec: &mut __wasi_ciovec_t) -> io::IoSliceMut {
+pub unsafe fn ciovec_to_host_mut(ciovec: &mut __wasi_ciovec_t) -> io::IoSliceMut {
```
Same here.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
 }
 
-pub(crate) unsafe fn iovec_to_host(iovec: &__wasi_iovec_t) -> io::IoSlice {
+pub unsafe fn iovec_to_host(iovec: &__wasi_iovec_t) -> io::IoSlice {
```
Same here.

kubkon commented on src/host.rs:
```diff
@@ -1,1049 +1,73 @@
 }
 
-pub(crate) unsafe fn iovec_to_host_mut(iovec: &mut __wasi_iovec_t) -> io::IoSliceMut {
+pub unsafe fn iovec_to_host_mut(iovec: &mut __wasi_iovec_t) -> io::IoSliceMut {
```
Same here.

kubkon commented on src/wasi.rs:
```diff
@@ -1,9 +1,6 @@
+//! fixed-size types, so it's host/target independent.
 
 #![allow(non_camel_case_types)]
 #![allow(non_snake_case)]
```
I believe `#![allow(dead_code)]` is no longer needed in the line below. Could you verify and remove it while we're here?

sunfishcode replied:
Removing it broke the Windows build, so I've reinstated it for now.

kubkon commented on src/wasi32.rs:
```diff
@@ -0,0 +1,232 @@
+
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+#![allow(dead_code)]
```
I believe `#![allow(dead_code)]` is no longer needed. Could you verify and remove it while we're here?

kubkon reviewed (approved):
LGTM, just a couple of minor nitpicks!

sunfishcode committed:
Use pub(crate).

kubkon commented:
And of course I forgot to check the Windows platform for the unused code. Perhaps in those 4 instances that caused the CI to fail on Windows:

```shell
error: constant item is never used: `RIGHTS_BLOCK_DEVICE_BASE`
   --> src\wasi.rs:152:1
    |
152 | pub(crate) const RIGHTS_BLOCK_DEVICE_BASE: __wasi_rights_t = RIGHTS_ALL;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D dead-code` implied by `-D warnings`

error: constant item is never used: `RIGHTS_BLOCK_DEVICE_INHERITING`
   --> src\wasi.rs:153:1
    |
153 | pub(crate) const RIGHTS_BLOCK_DEVICE_INHERITING: __wasi_rights_t = RIGHTS_ALL;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: constant item is never used: `RIGHTS_CHARACTER_DEVICE_BASE`
   --> src\wasi.rs:154:1
    |
154 | pub(crate) const RIGHTS_CHARACTER_DEVICE_BASE: __wasi_rights_t = RIGHTS_ALL;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: constant item is never used: `RIGHTS_CHARACTER_DEVICE_INHERITING`
   --> src\wasi.rs:155:1
    |
155 | pub(crate) const RIGHTS_CHARACTER_DEVICE_INHERITING: __wasi_rights_t = RIGHTS_ALL;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

we could add `#[allow(unused)]` next to each item? What are your thoughts on this? Ultimately, when we fill in the blanks in the Windows impl, all of the above constants will be used :-)

sunfishcode commented:
Oops, I didn't see your comment here. I'll clean that up in a followup PR.

kubkon commented:
> Oops, I didn't see your comment here. I'll clean that up in a followup PR.

Np! :-)
